File: 503.py
#!/usr/bin/env python

class Solution1(object):
	def nextGreaterElements(self, nums):
		numslen = len(nums)
		result = [-1] * numslen
		
		for idx in range(numslen):

			idx2 = (idx + 1) % numslen
			while idx2 != idx:
				if nums[idx] < nums[idx2]:
					result[idx] = nums[idx2]
					break

				idx2 = (idx2 + 1) % numslen

		return result

class Solution(object):
	def nextGreaterElements(self, nums):
		numslen = len(nums)
		result = [-1] * numslen
		stk = []
		nummap = dict()

		for idx1 in range(numslen):
			idx2 = (idx1 + 1) % numslen 
			while len(stk) != 0 and nums[stk[-1]] < nums[idx1] and idx2 != idx1:
				nummap[stk.pop()] = idx1
				idx2 = (idx2 + 1) % numslen

			stk.append(idx1)

		for idx1 in range(numslen):
			idx2 = (idx1 + 1) % numslen 
			while len(stk) != 0 and nums[stk[-1]] < nums[idx1] and idx2 != idx1:
				nummap[stk.pop()] = idx1
				idx2 = (idx2 + 1) % numslen

		for idx in range(numslen):
			if idx in nummap:
				result[idx] = nums[nummap[idx]]

		return result

if __name__ == '__main__':
	
	a = [1, 2, 1]
	sol = Solution()

	print sol.nextGreaterElements(a)

File: 503.cc
#include <iostream>
#include <vector>

using namespace std;

class Solution {

public:
	vector<int> nextGreaterElements(vector<int> &nums)
	{
		vector<int> results(nums.size(), -1);
		
		for(int i = 0; i<nums.size(); i++) {
			int j = (i + 1) % nums.size();
			
			while(j != i) {
				if(nums[i] < nums[j]) {
					results[i] = nums[j];
					break;
				}
				j = (j + 1) % nums.size();
			}
		}

		return results;
	}
};

int main(int argc, char **argv)
{
	Solution sol;
	vector<int> nums{1, 2, 1};

	vector<int> results = sol.nextGreaterElements(nums);

	for(auto &num: results) {
		cout<<num<<" ";
	}

	return 0;
}
